platforms/core-execution/snapshots/README.md

```diff
@@ -0,0 +1,168 @@
+
+We use [Merkle trees](https://en.wikipedia.org/wiki/Merkle_tree) to generate a single hash representing complex inputs.
+This involves hashing together hashes of individual components to create a hash for the whole.
+
+## Virtual File-System
+
+The execution engine regularly needs information about the state of the file-system in the form of snapshots taken of files and directories.
+It would be inefficient to re-read file hierarchies every time a snapshot of a part is needed, so file-system state is cached in the virtual file-system (VFS). 
+
+![Virtual file-system](Virtual%20File-System.drawio.svg)
+
+The VFS only stores information about the directory hierarchy of the running build.[^vfs-previous-builds]
+This means that we do not retain information about file-system objects outside the build directory.
+
+[^vfs-previous-builds]: Data about previous builds can also be retained for a while.
+
+The VFS is stored in an efficient sparse tree data structure using `FileSystemNode`s.
+This data-structure allows for reusing already taken snapshots when a parent directory's snapshot is 
+requested.
+It also supports storing filtered snapshots (e.g. when only `*.java` files are requested from a directory hierarchy).
+
+The VFS is not exposed directly; the entry point is via `FileSystemAccess`.
+This service has multiple `read()` methods to acquire file and directory snapshots.
+
+### Change Tracking
+
+For the VFS to correctly cache file-system state care must be taken to invalidate parts of the VFS before the file-system is modified.
+It is assumed that during a running build changes to file-system areas tracked by the VFS are always invalidated.
+So code modifying the file-system should either be wrapped in `FileSystemAccess.write()`, or a call to `FileSystemAccess.invalidate()` should be made before the changes are enacted.
+
+![File-system watching](File-System%20Watching.drawio.svg)
+
+#### File-System Watching
+
+To track modifications between builds, the VFS **watches the file-system** for changes on supported platforms.
+When a modification happens, any parts of the VFS data that might be out-of-date is discarded.
+
+When file-system watching is not available, we discard all VFS data collected during the build.
+
+The VFS and the execution engine in general does not distinguish between content available directly and content accessed through symlinks.
+However, file-system watching does not notify of changes happening to symlinked content reliably.
+Because of this we discard content accessed via symlinks from the VFS at the end of the build even if file-system watching is available.
+
+#### Caching File Hashes
+
+Actual file content hashing is handled by the `FileHasher`.
+File hashes are cached in-memory and on disk based on file modification dates and last modification times.
```

Comment:
What's the difference between "file modification date" and "last modification date"?